math/src/field/smallprimefield/mod.rs

```diff
@@ -0,0 +1,493 @@
+        let g_pow = div_by_power_of_two(M, Self::TWO_ADICITY);
+        let pow_val = G ^ g_pow;
+        Self::new(pow_val)
+    }
+
+    const fn get_modulus_bits() -> u32 {
+        64u32 - M.leading_zeros()
+    }
+
+    /// Creates a new field element from a u64 value. If the value is greater than or equal to
+    /// the field modulus, modular reduction is silently preformed. This function can also be used
+    /// to initialize constants.
+    pub const fn new(value: u64) -> Self {
+        let val = value % M;
+        Self { value: val }
+    }
+
+    #[allow(unused)]
+    pub unsafe fn get_power_series(b: Self, n: usize) -> Vec<Self> {
+        let mut result = utils::uninit_vector(n);
+        result[0] = Self::ONE;
+        for i in 1..result.len() {
+            result[i] = result[i - 1] * b;
+        }
+        result
+    }
+
+    pub unsafe fn get_twiddles(domain_size: usize) -> Vec<Self> {
+        debug_assert!(
+            domain_size.is_power_of_two(),
+            "domain size must be a power of 2"
+        );
+        let root = Self::get_root_of_unity(domain_size.try_into().unwrap());
+        let mut twiddles = Self::get_power_series(root, domain_size / 2);
+        fft::permute(&mut twiddles);
+        twiddles
+    }
+
+    pub unsafe fn get_inv_twiddles(domain_size: usize) -> Vec<Self> {
+        debug_assert!(
+            domain_size.is_power_of_two(),
+            "domain size must be a power of 2"
+        );
+        let root = Self::get_root_of_unity(domain_size.try_into().unwrap());
+        let inv_root = root.exp((domain_size as u32 - 1).into());
+        let mut inv_twiddles = Self::get_power_series(inv_root, domain_size / 2);
+        fft::permute(&mut inv_twiddles);
+        inv_twiddles
+    }
+}
+
+impl<const M: u64, const G: u64, const T: u32> FieldElement for BaseElement<M, G, T> {
+    type PositiveInteger = u64;
+    type BaseField = Self;
+
+    const ZERO: Self = Self::new(0);
+    const ONE: Self = Self::new(1);
+
+    const ELEMENT_BYTES: usize = ELEMENT_BYTES;
+
+    const IS_CANONICAL: bool = true;
+
+    fn inv(self) -> Self {
+        Self::new(inv(self.value, M))
+    }
+
+    fn conjugate(&self) -> Self {
+        Self::new(self.value)
+    }
+
+    fn elements_as_bytes(elements: &[Self]) -> &[u8] {
+        let len = elements.len() * Self::ELEMENT_BYTES;
+        let p = elements.as_ptr();
+        unsafe { slice::from_raw_parts(p as *const u8, len) }
+    }
+
+    unsafe fn bytes_as_elements(bytes: &[u8]) -> Result<&[Self], DeserializationError> {
+        if bytes.len() % Self::ELEMENT_BYTES != 0 {
+            return Err(DeserializationError::InvalidValue(format!(
+                "number of bytes ({}) does not divide into whole number of field elements",
+                bytes.len(),
+            )));
+        }
+
+        let p = bytes.as_ptr();
+        let len = bytes.len() / Self::ELEMENT_BYTES;
+
+        if (p as usize) % mem::align_of::<u128>() != 0 {
```

Comment:
Since the backing type is `u64`, we can probably relax the alignment a bit and make sure the values align on `u64` boundaries.